### PR TITLE
Add endpoint for fetching subscriber history

### PR DESCRIPTION
**Pull request by guarani**
### Description

Add endpoint for fetching subscriber history.

Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/23044

### Testing Details

- Check if `StatsSubscribersSummaryDataTests` passes

---

- [x] Please check here if your pull request includes additional test coverage.
- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            self.count = count
+        }
+    }
+
+    public init?(jsonDictionary: [String: AnyObject]) {
+        guard
+            let fields = jsonDictionary["fields"] as? [String],
+            let data = jsonDictionary["data"] as? [[Any]],
+            let dateIndex = fields.firstIndex(of: "period"),
+            let countIndex = fields.firstIndex(of: "subscribers")
```
Comment (staskus) [inline Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift:50]: This pattern of having an array of fields is in a way super-flexible but unfortunately incompatible with any benefits that `Codable` may provide.

File: Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                let date = StatsSubscribersSummaryData.dateFormatter.date(from: dateString),
+                let count = elements[countIndex] as? Int
+            else {
+                return nil
+            }
+
+            return SubscriberData(date: date, count: count)
+        }
+
+        let sorted = history.compactMap { $0 }.sorted(by: { $0.date.compare($1.date) == .orderedAscending })
```
Comment (staskus) [inline Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift]: nitpick

```suggestion
        let sorted = history.compactMap { $0 }.sorted { $0.date < $1.date }
```

File: Sources/WordPressKit/Services/StatsServiceRemoteV2.swift [hunk 3]
```diff
@@ -316,6 +316,30 @@ private extension StatsServiceRemoteV2 {
     }
 }
 
+// MARK: - Subscribers Data
+
+public extension StatsServiceRemoteV2 {
```
Comment (staskus) [inline Sources/WordPressKit/Services/StatsServiceRemoteV2.swift]: I think it's not necessary to create a separate method for subscribers endpoint.

I compared the API response and request with `stats/visits` and it looks to be following the same structure:

Same structure between subscribers and visits

What it means that it's enough for `StatsSubscribersSummaryData` to conform to `StatsTimeIntervalData` protocol, and follow the same rules when parsing as `StatsSummaryTimeIntervalData`. 

Then you could reuse `statsService.getData` method, which will also support quantity, end dates, `StatsPeriodUnit` type parameters out of the box. We won't need to be making changes to the API when adding additional date-picking capabilities to subscribers.

Example usage:

```
        statsService.getData(for: .day, endingOn: StatsDataHelper.currentDateForSite(), limit: 30) { (data: StatsSubscribersSummaryData?, error: Error?) in
            DispatchQueue.main.async {
                if let data = data {
                    cache.setValue(data, key: cacheKey)
                    self.chartSummary.send(.success(data))
                }
            else {
                    if cachedData == nil {
                        self.chartSummary.send(.error)
                    }
                }
            }
        }
```

I tested and it all works well. If you agree, feel free to cherry-pick this commit:

https://github.com/wordpress-mobile/WordPressKit-iOS/compare/task/subscribers-api-endpoint...task/subscribers-api-endpoint-review?expand=1

File: Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift (hunk 1 above)
Comment (guarani) [reply Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift:50]: I tried to use codable here with things like `nestedUnkeyedContainer` and `singleValueContainer` but kept running into issues. 

 I think leaving this as-is is fine for now., but please let me know if you happen to come across a way of using `Codable` 🙇

File: Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift (hunk 2 above)
Comment (guarani) [reply Sources/WordPressKit/Models/Stats/StatsSubscribersSummaryData.swift]: Makes sense 👍

File: Sources/WordPressKit/Services/StatsServiceRemoteV2.swift (hunk 3 above)
Comment (guarani) [reply Sources/WordPressKit/Services/StatsServiceRemoteV2.swift]: That's a very good idea! I went ahead and added it 👍

**Review by staskus — APPROVED** [review]
LGTM 👍

**Pull request merged**